[KEP] FlavorFungability: replace FlavorFungibilityImplicitPreferenceDefault feature gate with API (#7317)
kubernetes-sigs/kueue: keps/582-preempt-based-on-flavor-order/README.md

### Cluster Queue API

We extend the Cluster Queue API to introduce the new field `flavorFungibility` to opt-in and configure the behavior, and a minimal preference field for choosing borrowing-first vs preemption-first when the scheduler is exploring multiple flavors.

For each type of resource in each podSet, Kueue will traverse all resource groups and resource flavors to find an available flavor in present. When there are insufficient resources in the flavor, Kueue will prioritize preemption or borrowing based on the configured policy and, when applicable, the preference.

```
const (
    Borrow        FlavorFungibilityPolicy = "Borrow"        // deprecated alias of MayStopSearch
    Preempt       FlavorFungibilityPolicy = "Preempt"       // deprecated alias of MayStopSearch
    MayStopSearch FlavorFungibilityPolicy = "MayStopSearch"
    TryNextFlavor FlavorFungibilityPolicy = "TryNextFlavor"
)

> type FlavorAssignmentPreference string

Reviewer: ```suggestion
type FlavorFungibilityPreference string
```

I would recommend using `FlavorFungibility` or `FlavorFungibilityAssignment` prefix to avoid conflict in the future.

const (
    PreferenceBorrowing  FlavorAssignmentPreference = "Borrowing"

>     PreferencePreempting FlavorAssignmentPreference = "Preempting"

Reviewer 2: But if you prefer the verb form Im totally ok with that. Just let us know @tenzen-y, seems like there is no just one good answer

>     PreferencePreempting FlavorAssignmentPreference = "Preempting"

Reviewer: `Preemption`/`Borrowing` sounds good to me. Thanks.

  //
  // +kubebuilder:validation:Enum={MayStopSearch,TryNextFlavor,Borrow}
  // +kubebuilder:default="MayStopSearch"
  WhenCanBorrow FlavorFungibilityPolicy  `json:"whenCanBorrow,omitempty"`

  // whenCanPreempt determines whether a workload should try the next flavor
  // before preempting in current flavor. The possible values are:
  //
  // - `MayStopSearch`: stop the search for candidate flavors if workload fits or requires
  //   preemption to fit.
  // - `TryNextFlavor` (default): try next flavor if workload requires preemption
  //   to fit in current flavor.
  // - `Preempt` (deprecated): old name for `MayStopSearch`; please use new name.
  //
  // +kubebuilder:validation:Enum={MayStopSearch,TryNextFlavor,Preempt}
  // +kubebuilder:default="TryNextFlavor"
  WhenCanPreempt FlavorFungibilityPolicy `json:"whenCanPreempt,omitempty"`

  // preference selects the order between minimizing preemption and avoiding borrowing
  // when both WhenCanBorrow and WhenCanPreempt are TryNextFlavor.
  // If unset, the default preference is BorrowingOverPreemption.
  // Note: Preference=BorrowingOverPreemption does NOT mean we choose Borrowing over NoBorrowing;

>   // it only prefers Borrowing over Preempting among feasible outcomes.

Reviewer 2: ```suggestion
  // preference guides the choosing of the flavor for admission in case all candidate flavors
  // require either preemption, borrowing, or both. The possible values are:
  // - `BorrowingOverPreemption` (default): prefer to use borrowing rather than preemption
  // when such a choice is possible. More technically it minimizes the borrowing distance
  // in the cohort tree, and solves tie-breaks by preferring better preemption mode
  // (reclaim over preemption within ClusterQueue).
  // - `PreemptionOverBorrowing`: prefer to use preemption rather than borrowing
  // when such a choice is possible.  More technically it optimizes the preemption mode
  // (reclaim over preemption within ClusterQueue), and solves tie-breaks by minimizing
  // the borrowing distance in the cohort tree.
```

| `Fit`, `Borrow` | `WhenCanBorrow = MayStopSearch` | break |
| `Fit`, `NoBorrow`| any| break| 

By `Borrow`/`NoBorrow` we mean whether borrowing is required or not to fit the considered podset in the simulation result. After we complete the loop, either by trying all the flavors or by breaking out of it, we end up with a list of possible flavors containing all the considered flavors that yielded a result different than `NoFit`. We choose the flavor to assign based on the following default preference order: (`Fit`, `NoBorrow`), (`Fit`, `Borrow`), (`Preempt`, `NoBorrow`), (`Preempt`, `Borrow`).

When both `WhenCanBorrow` and `WhenCanPreempt` are `TryNextFlavor`, a ClusterQueue MAY set:
`spec.flavorFungibility.preference: BorrowingOverPreemption | PreemptionOverBorrowing`.

If preference is:
- BorrowingOverPreemption (or unset): (`Fit`, `NoBorrow`) > (`Fit`, `Borrow`) > (`Preempt`, `NoBorrow`) > (`Preempt`, `Borrow`).
- PreemptionOverBorrowing:            (`Fit`, `NoBorrow`) > (`Preempt`, `NoBorrow`) > (`Fit`, `Borrow`) > (`Preempt`, `Borrow`).

> The `PreemptionOverBorrowing` mode is introduced alongside the `FlavorFungibilityImplicitPreferenceDefault` feature gate; enabling that gate applies the new ordering without requiring each ClusterQueue to opt in.

Reviewer 2: ```suggestion
The `PreemptionOverBorrowing` mode is introduced as a replacement for the `FlavorFungibilityImplicitPreferenceDefault` feature gate; enabling that gate applies the new ordering without requiring each ClusterQueue to opt in.
```

[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
-->

The feature gate `FlavorFungibilityImplicitPreferenceDefault` was a temporary measure.

> Preference is now configured per ClusterQueue via `spec.flavorFungibility.preference`; the gate has been removed.

Reviewer: Additionally, does this include the deprecation of `FlavorFungibilityImplicitPreferenceDefault` FG?
If yes, could you add a Note as I proposed in another thread?

```suggestion
The feature gate `FlavorFungibilityImplicitPreferenceDefault` was supported between v0.13 and v0.14 and it was removed in v0.15.
Since v0.15, the Preference is now configured per ClusterQueue via `spec.flavorFungibility.preference`.
```

Reviewer 2: Yes the FG should be deprecated and removed in 0.17. The filed overrides the meaning of the FG until then. Please update

Author: 0.17 or 0.15? should I keep the feature gate then until 0.17?

[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
-->

> The feature gate `FlavorFungibilityImplicitPreferenceDefault` remains available through v0.16 to ease the transition to explicit preferences and will be removed in v0.17.

Reviewer 2: ```suggestion
The feature gate `FlavorFungibilityImplicitPreferenceDefault` is deprecated, but remains available through v0.16 to ease the transition to explicit preferences and will be removed in v0.17.
```
